src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -217,46 +171,152 @@ public void GarbageCollectInstalledWorkloadPacks(DirectoryPath? offlineCache = n
+                    Log?.LogMessage($"{e.StackTrace}");
+                    continue;
                 }
             }
-            catch (Exception e)
+        }
+
+        private void RemoveWorkloadPacks(List<WorkloadPackRecord> packsToRemove, DirectoryPath? offlineCache)
+        {
+            foreach (WorkloadPackRecord record in packsToRemove)
             {
-                LogException(e);
-                throw;
+                // We need to make sure the product is actually installed and that we're not dealing with an orphaned record, e.g.
+                // if a previous removal was interrupted. We can't safely clean up orphaned records because it's too expensive
+                // to query all installed components and determine the product codes associated with the component that
+                // created the record.
+                DetectState state = DetectPackage(record.ProductCode, out Version _);
+
+                if (state == DetectState.Present)
+                {
+                    // Manually construct the MSI payload package details
+                    string id = $"{record.MsiId}.Msi.{HostArchitecture}";
+                    MsiPayload msi = GetCachedMsiPayload(id, record.MsiNuGetVersion.ToString(), offlineCache);
+
+                    // Make sure the package we have in the cache matches with the record. If it doesn't, we'll do the uninstall
+                    // the hard way
+                    if (!string.Equals(record.ProductCode, msi.ProductCode, StringComparison.OrdinalIgnoreCase))
+                    {
+                        Log?.LogMessage($"ProductCode mismatch! Cached package: {msi.ProductCode}, pack record: {record.ProductCode}.");
+                        string logFile = GetMsiLogName(record, InstallAction.Uninstall);
+                        uint error = ExecuteWithProgress(String.Format(LocalizableStrings.MsiProgressUninstall, id), () => UninstallMsi(record.ProductCode, logFile));
+                        ExitOnError(error, $"Failed to uninstall {msi.MsiPath}.");
+                    }
+                    else
+                    {
+                        // No need to plan. We know that there are no other dependents, the MSI is installed and we
+                        // want to remove it.
+                        VerifyPackage(msi);
+                        ExecutePackage(msi, InstallAction.Uninstall);
+                    }
+                }
+            }
+        }
+
+        public void EradicateAllWorkloadInstallationRecords(SdkFeatureBand featureBand)
+        {
+            var allFeatureBands = RecordRepository.GetFeatureBandsWithInstallationRecords();
+
+            foreach (SdkFeatureBand potentialBandToClean in allFeatureBands)
+            {
+                if (potentialBandToClean.CompareTo(featureBand) < 1)
+                {
+                    var workloadInstallationRecordIds = RecordRepository.GetInstalledWorkloads(potentialBandToClean);
```

Comment:
This will delete records from VS as well which we don't want. Need to limit this

Reply:
Ack, I was hoping this would be done before I'm out tomorrow so it'd be reviewable. I think this is the last item remaining though, so pinging for review.

```suggestion
                    var workloadInstallationRecordIds = RecordRepository.GetCLIInstalledWorkloads(potentialBandToClean);
```

When I am back, I will implement `GetCLIInstalledWorkloads` and see if we can determine if the workload has a VS installation source in the registry key, so we don't return it and remove the VS workload installation record for MSI based clean --all.